lua/ttt2/extensions/draw.lua

```diff
@@ -457,37 +457,118 @@ function draw.AdvancedText(text, font, x, y, color, xalign, yalign, shadow, scal
+			-- overshoot the end of the word, so we need to put the rest
+			-- into the next line
+			if currentStartPos <= charCount then
+				lines[currentLineNumber] = fastutf8.sub(word, currentStartPos, charCount)
+			end
+
+			break
+		end
+
+		local nextLine = fastutf8.sub(word, currentStartPos, currentEndPos)
+		local widthNextLine = surface.GetTextSize(nextLine)
+
+		-- Check if our estimated cut needs adjustment and does not fit
+		if widthNextLine > width then
+			-- We need to keep removing characters until the line fits
+			local charsToRemove = 0
+			-- We keep track of the width of the removed chars
+			-- to not use the expensive utf8.sub function for each char
+			local widthOfRemovedChars = 0
+
+			-- Iterate from the end of the new line to the start
+			-- To never remove the last char, we add +1 to the start pos,
+			-- so we prevent infinite loops
+			for i = currentEndPos, currentStartPos + 1, -1 do
+				widthOfRemovedChars = widthOfRemovedChars + surface.GetTextSize(fastutf8.GetChar(word, i))
+				charsToRemove = charsToRemove + 1
+
+				if widthNextLine - widthOfRemovedChars <= width then
+					break
+				end
+			end
+
+			-- Only do something if we actually removed chars
+			if charsToRemove > 0 then
+				-- Remove the chars from the line & shift the next start position
+				nextStartPos = nextStartPos - charsToRemove
+				nextLine = fastutf8.sub(word, currentStartPos, nextStartPos - 1)
+			end
+		end
+
+		-- Add the line to the table
+		lines[currentLineNumber] = nextLine
+
+		-- Set the index to the new start position
+		currentStartPos = nextStartPos
+		currentLineNumber = currentLineNumber + 1
+	end
+
+	return lines
+end
+
 local function InternalGetWrappedText(text, width, scale)
 	-- Any wrapping required?
 	local w, h = surface.GetTextSize(text)
 
 	if w <= width then
-		return {text}, w, h -- Nope, but wrap in table for uniformity
+		return { text }, w, h -- Nope, but wrap in table for uniformity
 	end
 
-	local words = string.Explode(" ", text) -- No spaces means you're screwed
-	local lines = {""}
+	local words = string.Explode(" ", text)
+	local lines = {}
 
 	for i = 1, #words do
-		local wrd = words[i]
+		local word = words[i]
 
-		if i == 1 then
-			-- add the first word whether or not it matches the size to prevent
-			-- weird empty first lines and ' ' in front of the first line
-			lines[1] = wrd
+		-- first, check the length of the word; if it is longer than a line, then
+		-- it has to be split as well
+		local widthWord = surface.GetTextSize(word)
+
+		if widthWord > width then
+			table.Add(lines, InternalSplitLongWord(word, width, widthWord))
 
 			continue
 		end
 
-		local lns = #lines
-		local added = lines[lns] .. " " .. wrd
+		local amountLines = #lines
+		local combinedString = ""
+
+		if i == 1 then
+			combinedString = word
+		else
+			combinedString = lines[amountLines] .. " " .. word
+		end
 
-		w = surface.GetTextSize(added)
+		w = surface.GetTextSize(combinedString)
 
 		if w > width then
```

Comment:
rename `w` to `widthCombinedString` or `widthCurrentLine` ? Feels wrong to make every var expressive but leave this out.